pcap-dpdk.c

```diff
@@ -0,0 +1,800 @@
+	u_char *cb_arg = pd->cb_arg;
+	int nb_rx=0;
+	struct rte_mbuf *pkts_burst[MAX_PKT_BURST];
+	struct rte_mbuf *m;
+	struct pcap_pkthdr pcap_header;
+	uint16_t portid = pd->portid;
+	unsigned lcore_id = rte_lcore_id();
+	unsigned master_lcore_id = rte_get_master_lcore();
+	// In DPDK, pkt_len is sum of lengths for all segments. And data_len is for one segment
+	uint16_t data_len = 0;
+	uint32_t pkt_len = 0;
+	int caplen = 0;
+	u_char *bp = NULL;
+	int i=0;
+	unsigned int gather_len =0;
+	int pkt_cnt = 0;
+	int is_accepted=0;
+	u_char *large_buffer=NULL;
+		
+	if(lcore_id == master_lcore_id){
+		RTE_LOG(DEBUG, USER1, "dpdk: lcoreid=%u runs for portid=%u\n", lcore_id, portid);
+	}else{
+		RTE_LOG(DEBUG, USER1, "dpdk: lcore %u has nothing to do\n", lcore_id);
+	}
+	//only use master lcore
+	if (lcore_id != master_lcore_id){
+		return;
+	}
+	if (max_cnt>0 && max_cnt < MAX_PKT_BURST){
+		burst_cnt = max_cnt;
+	}else{
+		burst_cnt = MAX_PKT_BURST;
+	}
+	while( max_cnt==-1 || pkt_cnt < max_cnt){
+		if (p->break_loop){
+			break;
+		}
+		nb_rx = (int)rte_eth_rx_burst(portid, 0, pkts_burst, burst_cnt);
+		pkt_cnt += nb_rx;
+		for ( i = 0; i < nb_rx; i++) {
+			m = pkts_burst[i];
+			calculate_timestamp(&(pd->ts_helper),&(pcap_header.ts));
+			data_len = rte_pktmbuf_data_len(m);
+			pkt_len = rte_pktmbuf_pkt_len(m);
+			// caplen = min(pkt_len, p->snapshot);
+			// caplen will not be changed, no matter how long the rte_pktmbuf
+			caplen = pkt_len < p->snapshot ? pkt_len: p->snapshot; 
+			pcap_header.caplen = caplen;
+			pcap_header.len = pkt_len; 
+			// volatile prefetch
+			rte_prefetch0(rte_pktmbuf_mtod(m, void *));
+			bp = NULL;
+			if (m->nb_segs == 1)
+			{
+				bp = rte_pktmbuf_mtod(m, u_char *);
+			}else{
+				// use fast buffer pcap_tmp_buf if pkt_len is small, no need to call malloc and free
+				if ( pkt_len <= ETHER_MAX_JUMBO_FRAME_LEN)
+				{
+					gather_len = dpdk_gather_data(pd->pcap_tmp_buf, RTE_ETH_PCAP_SNAPLEN, m);
+					bp = pd->pcap_tmp_buf;
+				}else{ 
+					// need call free later
+					large_buffer = (u_char *)malloc(caplen*sizeof(u_char));
+					gather_len = dpdk_gather_data(large_buffer, caplen, m); 
+					bp = large_buffer;
+				}
+				
+			}
+			if (bp){
+				//default accpet all
+				is_accepted=1;
+				if (pd->filter_in_userland && p->fcode.bf_insns!=NULL)
+				{
+					if (!pcap_filter(p->fcode.bf_insns, bp, pcap_header.len, pcap_header.caplen)){
+						//rejected
+						is_accepted=0;
+					}
+				}
+				if (is_accepted){
+					cb(cb_arg, &pcap_header, bp);
+				}else{
+					pd->bpf_drop++;
+				}
+			}
+			//free all pktmbuf
+			rte_pktmbuf_free(m);
+			if (large_buffer){
+				free(large_buffer);
+				large_buffer=NULL;
+			}
+		}
+	}	
+	pd->rx_pkts = pkt_cnt;
+}
+
+static int pcap_dpdk_dispatch(pcap_t *p, int max_cnt, pcap_handler cb, u_char *pcap_user)
+{
+	unsigned lcore_id = 0;	
+	struct pcap_dpdk *pd = (struct pcap_dpdk*)(p->priv);
+	pd->rx_pkts=0;
+	pd->cb = cb;
+	pd->cb_arg = pcap_user;
+	pd->max_cnt = max_cnt;
```

Comment:
Why not just pass `max_cnt` as an argument to `dpdk_dispatch_internal()`?

Reply:
Yes. The code is simple and clean.